### PR TITLE
Register marks

### PR DESCRIPTION
**Pull request by dfreeman06**
[binder](https://mybinder.org/v2/gh/jupyrdf/ipyelk/register_marks?urlpath=lab%2Ftree%2Fexamples%2F100_node_label_placement.ipynb )

Adds mapping of nodes in original graph to elk marks so hover and selection traitlets are in terms of the original nodes rather than elk identifiers. 

And currently still a work in progress for creating a text measurement widget to allow more accurate measuring of label sizes when passing off to the layout.

### CODE REVIEW

File: py_src/ipyelk/app.py [hunk 1]
```diff
@@ -10,6 +10,7 @@
 from .diagram import ElkDiagram, ElkLabel, ElkNode
 from .schema import ElkSchemaValidator
 from .styled_widget import StyledWidget
+from .toolbar import Toolbar
```
Comment (nrbgt) [inline py_src/ipyelk/app.py:13]: :+1:

File: py_src/ipyelk/diagram/elk_text_sizer.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import logging
+
+# Copyright (c) 2020 Dane Freeman.
+# Distributed under the terms of the Modified BSD License.
+import traitlets as T
+from ipywidgets import DOMWidget
+
+from .._version import EXTENSION_NAME, EXTENSION_SPEC_VERSION
+
+logger = logging.getLogger()
```
Comment (nrbgt) [inline py_src/ipyelk/diagram/elk_text_sizer.py:13]: all widgets have `self.log`

File: py_src/ipyelk/diagram/elk_text_sizer.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    text = T.Unicode().tag(sync=True)
+    width = T.Float().tag(sync=True)
+    height = T.Float().tag(sync=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.futures = {}
+        self.on_msg(self._resolve_measurement)
+        self.queue = asyncio.Queue()
```
Comment (nrbgt) [inline py_src/ipyelk/diagram/elk_text_sizer.py:34]: signal private as `_queue`?

File: py_src/ipyelk/nx/transformer.py [hunk 4]
```diff
@@ -146,6 +166,20 @@ def size_nodes(self, nodes: Dict[Hashable, ElkNode]) -> Dict[Hashable, ElkNode]:
...
+        # TODO improved nested dictionary merging
+
+        overrides = self.mark_overrides.get(item, {})
+        properties = overrides.get("properties", {})
+        css_classes = properties.get("cssClasses", [])
+        if styles:
+            css_classes += styles
+
+        properties = dict(
+            **ChainMap(dict(cssClasses=" ".join(css_classes)), properties)
```
Comment (nrbgt) [inline py_src/ipyelk/nx/transformer.py:179]: :rofl:

File: py_src/ipyelk/diagram/elk_text_sizer.py (hunk 3 above)
Comment (dfreeman06) [reply py_src/ipyelk/diagram/elk_text_sizer.py:34]: Yes definitely need that private. Also the line above is the line that isn't working at the moment...  I am not getting the measurement message back at the self._resolve_measurement function.

File: py_src/ipyelk/nx/transformer.py (hunk 4 above)
Comment (dfreeman06) [reply py_src/ipyelk/nx/transformer.py:179]: it makes me feel so dirty

File: py_src/ipyelk/toolbar.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import ipywidgets as W
+import traitlets as T
+
+from .styled_widget import StyledWidget
+
+
+class Toolbar(W.HBox, StyledWidget):
```
Comment (nrbgt) [inline py_src/ipyelk/toolbar.py:10]: personally would like an option for this to be vertical....

File: py_src/ipyelk/tools/tools.py [hunk 6]
```diff
@@ -26,12 +26,21 @@ def _default_description(self):
...
                 self.app.refresh()
+
+
+class FitBtn(ToolButton):
+    @T.default("description")
+    def _default_description(self):
+        return "Fit"
+
+    def handler(self, *args):
+        diagram: ElkDiagram = self.app.diagram
```
Comment (nrbgt) [inline py_src/ipyelk/tools/tools.py:45]: inline annotations have a smell... it is _possible_ to hoist traitlets with e.g.
```
class Foo(HasTraits):
  bar = Unicode(allow_none=True)  # type: Optional[Text]
```

if we're going to be serious about this, we need to add e.g. mypy...

File: src/measure_text.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    super.initialize(parameters);
+    // this.model.on('change:text', this.measure, this);
+    console.log('init');
+    this.model.on('msg:custom', this.measure, this);
+
+    this.el = document.body.appendChild(document.createElement('div'));
+    this.el.classList.add('p-Widget');
+    this.el.classList.add(WIDGET_CLASS);
+
+    this.el.innerHTML = '<svg><g><g><text class="elklabel"></text></g></g></svg>';
```
Comment (nrbgt) [inline src/measure_text.ts:54]: assuming `elklabel` is a magic class provided by the upstream, should probably be hoisted to some common variables, e.g.
```ts
export const ELK_CSS = {
  label: 'elklabel',
  ...
}
```

File: src/measure_text.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    // this.model.set('height', size.height);
+    // this.touch();
+
+    var message = {
+      event: 'measurement',
+      text: value,
+      width: size.width,
+      height: size.height
+    };
+    console.log(message);
```
Comment (nrbgt) [inline src/measure_text.ts:80]: there is an `ELK_DEBUG` someplace...

File: style/index.css [hunk 9]
```diff
@@ -128,3 +128,37 @@
...
+  visibility: visible;
+  opacity: 0.25;
+  transform: translateY(0);
+}
+
+.jp-ElkApp:hover .jp-ElkToolbar:hover {
+  opacity: 1;
+}
+
+.jp-ElkToolbar .close-btn {
```
Comment (nrbgt) [inline style/index.css:152]: why the special case?

File: py_src/ipyelk/nx/transformer.py [hunk 10]
```diff
@@ -389,6 +463,20 @@ def closest_common_visible(self, nodes: Tuple[Hashable]) -> Hashable:
         result = lowest_common_ancestor(tree, [self.closest_visible(n) for n in nodes])
         return result
 
+    def from_id(self, element_id: str) -> Hashable:
+        """Use the elk identifiers to find original objects"""
+        try:
+            return self._elk_to_item[element_id]
+        except KeyError:
+            raise ValueError(f"Element id `{element_id}` not in elk id registry.")
```
Comment (nrbgt) [inline py_src/ipyelk/nx/transformer.py:471]: perhaps make some `ElkError` subclass... built-in exception types are harder to reason about.

File: py_src/ipyelk/toolbar.py (hunk 5 above)
Comment (nrbgt) [reply py_src/ipyelk/toolbar.py:10]: Will be discussed further on #22

**Comment by nrbgt** [comment]
It's probably appropriate to go ahead and bump the version to `0.2.0` already on this branch, as it changes APIs.

The widget JSON spec number might also need to be bumped...

**Comment by nrbgt** [comment]
So, do the `bulk_sizer.futures` ever get cleaned up? It seems like one could easily get a _lot_ in here. Perhaps a lighter-weight cache of `{(text, options): [width, height]}`?

Also, the placement widget is huge (which is fine)... perhaps the toolbar (before or after #22) might offer a cut-down placement tool?

```
|-------|      |---------------------------|  |---------------------------|
| [fit] | ...  | [align right|center|left] |  | [align top|middle|bottom] |
|-------|      |---------------------------|  |---------------------------|
```

File: py_src/ipyelk/nx/transformer.py (hunk 4 above)
Comment (sanbales) [reply py_src/ipyelk/nx/transformer.py:193]: Dane: `ChainMap`'s #1 Fan  :-)

File: py_src/ipyelk/nx/transformer.py [hunk 11]
```diff
@@ -167,35 +200,76 @@ def process_edges(
...
         )
-
-    def make_port(self, owner, port, styles):
+        self.register(elk_edge, edge)
+        return elk_edge
+
+    def make_port(self, owner: Hashable, port: Hashable, styles: List[str]) -> ElkPort:
+        """Make the associated elk port for the given owner node and port
+
+        :param owner: [description]
```
Comment (sanbales) [inline py_src/ipyelk/nx/transformer.py:248]: could we fill in these descriptions?

File: py_src/ipyelk/app.py [hunk 12]
```diff
@@ -44,19 +45,31 @@ def refresh(self, change: T.Bunch = None) -> Dict:
 
         return self.value
 
+    def from_id(self, element_id: str) -> Hashable:
+        """Use the elk identifiers to find original objects"""
+        return element_id
+
+    def to_id(self, item: Hashable) -> str:
+        """Use original objects to find elk id"""
+        return item
```
Comment (sanbales) [inline py_src/ipyelk/app.py:66]: should this be `return str(item)`?

**Comment by nrbgt** [comment]
Push d178050
- bumped the version to `0.2.0` 
- updated the examples to have links to the new notebooks

File: py_src/ipyelk/diagram/elk_text_sizer.py (hunk 2 above)
Comment (nrbgt) [reply py_src/ipyelk/diagram/elk_text_sizer.py]: added a note about this in contributing, but haven't looked for stuff yet... something like [semgrep](https://github.com/returntocorp/semgrep) would be rad...

**Comment by nrbgt** [comment]
3a6df6e a little profile-guided optimization found some more situations where thousands of functions were being allocated...

**Review by nrbgt — APPROVED** [review]

File: py_src/ipyelk/tools/tools.py (hunk 6 above)
Comment (dfreeman06) [reply py_src/ipyelk/tools/tools.py:45]: Added followon issue https://github.com/jupyrdf/ipyelk/issues/25

File: anaconda-project-lock.yml [hunk 13]
```diff
@@ -19,221 +19,240 @@ env_specs:
...
+        - appdirs=1.4.4=pyh9f0ad1d_0
+        - async-lru=1.0.2=py_0
+        - async_generator=1.10=py_0
+        - attrs=20.2.0=pyh9f0ad1d_0
         - backcall=0.2.0=pyh9f0ad1d_0
         - backports.functools_lru_cache=1.6.1=py_0
         - backports=1.0=py_2
-        - black=19.10b0=py_4
-        - bleach=3.1.5=pyh9f0ad1d_0
+        - black=20.8b1=py_1
```
Comment (nrbgt) [inline anaconda-project-lock.yml:36]: We can roll this back if we don't want to deal with it _right_ right now...

File: py_src/ipyelk/diagram/elk_text_sizer.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+from dataclasses import dataclass, field
+from typing import Dict, List
+
+import traitlets as T
+from async_lru import alru_cache
+from ipywidgets import DOMWidget
+
+from .._version import EXTENSION_NAME, EXTENSION_SPEC_VERSION
+
+print("TODO: use alru_cache", alru_cache)
```
Comment (nrbgt) [inline py_src/ipyelk/diagram/elk_text_sizer.py:15]: over to you, @dfreeman06 !

File: py_src/ipyelk/nx/transformer.py [hunk 15]
```diff
@@ -235,23 +324,34 @@ def get_node_size(self, node: ElkNode) -> Tuple[Optional[float], Optional[float]
...
-        ]
+        size = await self.size_label(name)
+        # width = self.sizer.measure(name)
+        label = ElkLabel(
+            id=f"{name}_label_{node}",
+            text=name,
+            width=size.width,
+            height=size.height,
+            # x=self.label_offset,
+            # y=self.label_offset,
```
Comment (nrbgt) [inline py_src/ipyelk/nx/transformer.py:350]: these staying in?

File: atest/_resources/keywords/Lab.robot [hunk 16]
```diff
@@ -206,7 +206,7 @@ Restart and Run All
     Lab Command    Restart Kernel and Run All Cells
     Accept Default Dialog Option
     Ensure Sidebar Is Closed
-    Wait Until Element Contains    ${JLAB XP LAST CODE PROMPT}    [*]:
+    Run Keyword and Ignore Error    Wait Until Element Contains    ${JLAB XP LAST CODE PROMPT}    [*]:
```
Comment (nrbgt) [inline atest/_resources/keywords/Lab.robot:209]: this seems to be flaky on windows... might have to be a "wait until keyword succeeds" kind of thing...

**Comment by nrbgt** [comment]
I've tried slowing down windows CI some... part of the problem is that we're still on lab 1, which can't help but try to do crazy nodejs stuff to see if it needs to build itself, and node is not kind there.

**Comment by nrbgt** [comment]
Cool, actually got a windows fail on robot... and better still, it fixed itself on later push!

```
8 critical tests, 7 passed, 1 failed
8 tests total, 7 passed, 1 failed

===================================================

Output:  D:\a\ipyelk\ipyelk\atest\output\windows_3\output.xml
XUnit:   D:\a\ipyelk\ipyelk\atest\output\windows_3\xunit.xml
Log:     D:\a\ipyelk\ipyelk\atest\output\windows_3\log.html
Report:  D:\a\ipyelk\ipyelk\atest\output\windows_3\report.html
Total testing: 21 minutes 40.89 seconds

Elapsed time:  3 minutes 5.0 seconds
main exited with 1

1 errors in 185 seconds
DependencyError - taskid:all
ERROR: Task 'all' checking dependencies: Dependent file 'D:\a\ipyelk\ipyelk\build\robot.windows_success.ok' does not exist.
```

**Comment by nrbgt** [comment]
Approval remains... I didn't quite get to doing the text sizer testing, and am likely not going to get anywhere else this week, so that should not block.  I would recommend merge-and-iterate as we look at other things that need testing next week, but we'll want these changes in other branches soon.

File: py_src/ipyelk/nx/transformer.py (hunk 15 above)
Comment (dfreeman06) [reply py_src/ipyelk/nx/transformer.py:342]: I'll take a look at these on https://github.com/jupyrdf/ipyelk/issues/19.

**Pull request merged**